### DISCUSSION TITLE
Editing the esxi_decoders.xml decoder

### DISCUSSION THREAD

**Discussion by Kobrik1** [wazuh/wazuh#23786] [Q&A] [answered]
Hi, I am trying to get esxi_decoders.xml and esxi_rules.xml working in Wazuch. The source of information I use is here:

[https://wazuh.com/blog/monitoring-vmware-esxi-with-wazuh](url)

The syslog from both ESXi and VCenter is routed directly to the wazuh server. So the logs are not routed to the Ubuntu endpoint, but directly. So the log looks like this:

> 2024-05-30T13:00:09. 957Z virt-esx-int-22 sandboxd[2099017]: --> [context]zKq7AVICAgAAAP////8Jc2FuZGJveGQAAANZQWxpYnZtYWNvcmUuc28AAI4PNwCieicA0HsnAI4xKACwTCgAS9lJAYJ6AGxpYnB0aHJlYWQuc28uMAAC7y4PbGliYy5zby42AA==[/context]

Instead of the expected start of the log: **vmware-esxi:**
so I need to modify the pre-decoder.
Instead of :
```
<decoder name="vmware-esxi">
  <prematch>^vmware-esxi: </prematch>
</decoder>
```

I would like to use something along the lines of:
```
<decoder name="vmware-esxi">
  <prematch>^\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d.\d\d\dZ</prematch>
</decoder>

```
Unfortunately, this doesn't work. What am I doing wrong?
Thank you very much for your help.

**Comment by Hasitha9796** [comment, marked as answer]
Hi [Kobrik1](https://github.com/Kobrik1)

I have noticed that in the predecoding section log timestamp is taken incorrectly.

<img width="222" alt="image" src="https://github.com/user-attachments/assets/5d0d1a17-7681-4722-b2fd-2f972a9e86c7" />

However, for this, you can use out_format in log collection.
https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/localfile.html#out-format

So you can use `<out_format>` tag to add an extra header to your logs and decode from that point. Need to replace the `/var/log/test.log` with your ESXI log file path.

```
 <localfile>
    <log_format>syslog</log_format>
    <location>/var/log/test.log</location>
    <out_format>ESXi: $(log)</out_format>
  </localfile>
```

After adding an extra header to your log, Wazuh Manager receives the full log below.
```ESXi: 2024-05-30T13:00:09. 957Z virt-esx-int-22 sandboxd[2099017]: --> [context]zKq7AVICAgAAAP////8Jc2FuZGJveGQAAANZQWxpYnZtYWNvcmUuc28AAI4PNwCieicA0HsnAI4xKACwTCgAS9lJAYJ6AGxpYnB0aHJlYWQuc28uMAAC7y4PbGliYy5zby42AA==[/context]```

Then you can create decoders based on that. I have created a few decoders to capture the relevant field values.
You can add these decoders into the custom decoder creation file: `/var/ossec/etc/decoders/local_decoder.xml`

```
<decoder name="ESXI">
    <prematch>ESXi: \d+-\d+-\S+</prematch>
</decoder>

<decoder name="ESKI-custom">
<parent>ESXI</parent>
<regex>(\d+-\d+-\S+\s\S+)\s(\S+)\s(\S+)[</regex>
<order>log_time,host,program</order>
</decoder>

<decoder name="ESKI-custom">
<parent>ESXI</parent>
<regex>[context](\.+)[/context]</regex>
<order>context</order>
</decoder>
```

Then, restart the Wazuh manager to apply changes.
`systemctl restart wazuh-manager`

Check these documents to learn more about regex and decoder syntax
[Decoders Syntax](https://documentation.wazuh.com/current/user-manual/ruleset/ruleset-xml-syntax/decoders.html)
[Regular Expression Syntax](https://documentation.wazuh.com/current/user-manual/ruleset/ruleset-xml-syntax/regex.html)

Let me know if you need further assistance on this.
